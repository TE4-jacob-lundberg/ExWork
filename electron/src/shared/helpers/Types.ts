export interface ILink {
  url: string;
  label: string;
}

export interface IGameFormFields {
  id: string;
  title: string;
  abbreviation: string;
  image: string; 
  xAxis: string;
  yAxis: string;
  links: ILink[];
  fileNames: string[];
}
