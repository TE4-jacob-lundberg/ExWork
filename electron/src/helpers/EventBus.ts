/* eslint-disable @typescript-eslint/no-explicit-any */
const EventBus: IEventBus = {
  events: {},
  dispatch: function (event: string, data: any): void {
    if (!this.events[event]) return;
    this.events[event].forEach((callBack: Function) => callBack(data));
  },
  subscribe: function (event: string, callBack: Function): void {
    if (!this.events[event]) this.events[event] = [];
    this.events[event].push(callBack);
  },
};

interface IEventBus {
  events: {[id: string]: Function[]};
  dispatch: (event: string, data: any) => void;
  subscribe: (event: string, callBack: Function) => void;
}

export { EventBus };
