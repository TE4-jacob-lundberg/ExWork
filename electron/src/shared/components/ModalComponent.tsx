import React from 'react';
import Styled from '@emotion/styled';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { css, jsx } from '@emotion/core';

/* @jsx jsx */

type ActionProps = {
  background: string;
}

interface Props {
  title: string;
  body: string;
  onClose: () => void;
  cancelAble?: boolean;
  cancelLabel?: string;
  onCancel: () => void;
  continueLabel?: string; 
  onContinue: () => void;
}

const ModalComponent: React.FC<Props> = function (props: Props) { 
  const ContainerStyled = Styled.div`
    position: absolute;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    background: #333;
    z-index: 999;
    height: 500px;
    width: 500px;
    padding-bottom: 8px;
  `;

  const TitleStyled = Styled.h1`
    width: 100%;
    height: 100px;
    background: #444;
    display: flex;
    align-items: center;
    justify-content: center;
  `;

  const BodyStyled = Styled.p`
    width: 80%;
    margin-top: 16px;
    text-align: center;
    margin: auto;
    font-size: 1.5rem;
  `;

  const ActionsContainerStyled = Styled.div`
    width: 100%;
    height: 50px;
    display: flex;
    justify-content: space-between;
    flex-direction: ${!props.cancelAble ? 'row-reverse' : 'row'};
    padding: 0 16px;
    align-items: center;
  `;

  const ActionStyled = Styled.button<ActionProps>`
    height: 80%;
    width: fit-content;
    padding: 0 8px;
    border: none;
    color: #FFF;
    font-size: 1.25rem;
    background: ${(props): string => props.background}
  `;

  return (
    <ContainerStyled className="Modal">
      <TitleStyled>{props.title}</TitleStyled>
      <i 
        css={css`
          position: absolute;
          top: 8px;
          right: 8px;
        `} 
        className="material-icons"
        data-testid="close-modal"
        onClick={(): void => props.onClose()}
      >
        close
      </i>
      <BodyStyled>{props.body}</BodyStyled>
      <ActionsContainerStyled>
        {props.cancelAble && <ActionStyled 
          onClick={(): void => props.onCancel()}
          data-testid="cancel-modal"
          background="red"
        >
          {props.cancelLabel || 'Cancel'}
        </ActionStyled>}
        <ActionStyled 
          onClick={(): void => props.onContinue()}
          data-testid="continue-modal"
          background="green"
        >
          {props.continueLabel || 'Continue'}
        </ActionStyled>
      </ActionsContainerStyled>
    </ContainerStyled>
  );
};

export default ModalComponent;
