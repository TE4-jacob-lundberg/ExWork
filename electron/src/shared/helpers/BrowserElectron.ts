// eslint-disable-next-line @typescript-eslint/no-explicit-any
export let electron: any;

if (process.env.ENVIRONMENT === 'test') electron = require('electron');
else electron = window.require('electron');
