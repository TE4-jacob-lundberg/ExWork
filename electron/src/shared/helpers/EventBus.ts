/* eslint-disable @typescript-eslint/no-explicit-any */
interface IEventBus {
  events: {[id: string]: Function[]};
  dispatch: (event: string, data: any) => void;
  subscribe: (event: string, callBack: Function) => void;
}

const EventBus: IEventBus = {
  events: {},
  dispatch: function (event: string, data: any): void {
    if (!this.events[event]) return;
    this.events[event].forEach((callBack: Function) => callBack(data));
  },
  subscribe: function (event: string, callBack: Function): void {
    if (!this.events[event]) this.events[event] = [];
    this.events[event].push(callBack);
  },
};

export { EventBus };
