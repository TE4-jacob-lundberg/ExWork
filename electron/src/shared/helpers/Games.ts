import { standardGames, IGame } from '../constants/standardGames';

export class Games {

  public static all(): IGame[] {
    if (!localStorage.getItem('games')) localStorage.setItem('games', JSON.stringify(standardGames));

    return JSON.parse(localStorage.getItem('games')!) || [];
  }

  public static addNew(game: IGame): string {
    const savedGames: IGame[] = JSON.parse(localStorage.getItem('games')!);
    if (savedGames.some(sGame => sGame.id === game.id)) return 'DUPLICATE_ID';
    
    localStorage.setItem('games', JSON.stringify(savedGames.concat(game)));
    return '';
  }
}
