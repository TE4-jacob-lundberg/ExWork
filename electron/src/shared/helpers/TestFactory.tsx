/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import { render, RenderResult } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { createBrowserHistory } from 'history';

export function factory(component: JSX.Element): RenderResult {
  return render(component);
}

export function routerFactory(component: JSX.Element, options: any = { history: createBrowserHistory() }): RenderResult {
  return factory(
    <Router history={options.history}>
      {component}
    </Router>
  );
}
