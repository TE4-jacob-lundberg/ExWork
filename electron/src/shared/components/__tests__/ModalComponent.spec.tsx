import React from 'react';
import { fireEvent } from '@testing-library/react';

import { factory } from '../../helpers/TestFactory';
import ModalComponent from '../ModalComponent';

const mockedClose = jest.fn();
const mockedCancel = jest.fn();
const mockedContinue = jest.fn();

const testComponent = <ModalComponent 
  title="Test"
  cancelAble
  onClose={mockedClose}
  onCancel={mockedCancel}
  onContinue={mockedContinue}
>
  text
</ModalComponent>;

describe('ModalComponent.tsx', () => {
  it('renders correctly', () => {
    const { container, getByText } = factory(testComponent);
    
    expect(getByText('Test'));
    expect(getByText('text'));
    expect(container).toMatchSnapshot();
  });

  it('closes when clicking on x', () => {
    const { getByTestId } = factory(testComponent);

    fireEvent.click(getByTestId('close-modal'));

    expect(mockedClose).toHaveBeenCalledTimes(1);
  });

  it('closes when clicking cancel', () => {
    const { getByTestId } = factory(testComponent);

    fireEvent.click(getByTestId('cancel-modal'));

    expect(mockedCancel).toHaveBeenCalledTimes(1);
  });

  it('proceeds when continue is clicked', () => {
    const { getByTestId } = factory(testComponent);

    fireEvent.click(getByTestId('continue-modal'));

    expect(mockedContinue).toHaveBeenCalledTimes(1);
  });
});
