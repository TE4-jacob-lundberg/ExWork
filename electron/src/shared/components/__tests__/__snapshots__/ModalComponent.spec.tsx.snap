// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalComponent.tsx renders correctly 1`] = `
<div>
  <div
    class="Modal css-1erj6aa"
  >
    <h1
      class="css-17mgc3o"
    >
      Test
    </h1>
    <i
      class="material-icons css-52vivv-ModalComponent"
      data-testid="close-modal"
    >
      close
    </i>
    <p
      class="css-11p02fo"
    >
      text
    </p>
    <div
      class="css-3to2p8"
    >
      <button
        class="css-1mir8w9"
        data-testid="cancel-modal"
      >
        Cancel
      </button>
      <button
        class="css-a4emjt"
        data-testid="continue-modal"
      >
        Continue
      </button>
    </div>
  </div>
</div>
`;
